src/shell-app.c

```diff
@@ -927,13 +928,45 @@ shell_app_on_user_time_changed (MetaWindow *window,
     }
 }
 
+static void
+shell_app_do_quit (ShellApp *app)
+{
+  GSList *iter;
+
+  if (app->running_state == NULL)
+    return;
+
+  for (iter = app->running_state->windows; iter; iter = iter->next)
+    {
+      MetaWindow *win = iter->data;
+
+      if (!shell_window_tracker_is_window_interesting (win))
+        continue;
+
+      meta_window_delete (win, shell_global_get_current_time (shell_global_get ()));
+    }
+}
+
 static void
 shell_app_sync_running_state (ShellApp *app)
 {
   g_return_if_fail (app->running_state != NULL);
 
   if (app->running_state->interesting_windows > 0)
     {
+      if (app->showing_splash && app->state == SHELL_APP_STATE_STOPPED) {
```

Comment:
I'm not entirely sure why this code here doesn't already accomplish this: https://github.com/endlessm/gnome-shell/blob/master/js/ui/appActivation.js#L284

It seems to me that we're trying to solve a race condition between the `SplashClosed` signal and how this class uses the speedwagon window count to determine the window state. A better fix may be to simply remove the `SplashClosed` signal altogether then, and only rely on the window manager notifications to determine that, if the splash screen was indeed closed before the app became running, we should quit it.

Reply:
> I'm not entirely sure why this code here doesn't already accomplish this: https://github.com/endlessm/gnome-shell/blob/master/js/ui/appActivation.js#L284

That doesn't work right now since the app changes it state to STOPPED before the `SplashClosed` D-Bus signal sent by the speedwagon process reaches the shell, so the [code that is supposed to set `this._abort = true'](https://github.com/endlessm/gnome-shell/blob/master/js/ui/appActivation.js#L157) has not been called at this point, resulting in `this._app.request_quit()` never being executed. And even if got executed, `request_quit()` would be a no-op since no windows associated to the app are yet created at this point (just the speedwagon).

> It seems to me that we're trying to solve a race condition between the `SplashClosed` signal and how this class uses the speedwagon window count to determine the window state. A better fix may be to simply remove the `SplashClosed` signal altogether then, and only rely on the window manager notifications to determine that, if the splash screen was indeed closed before the app became running, we should quit it.

Yes, indeed there's a race condition and I also thought of removing that signal, but was not 100% sure about missing some use cases. I'll give it another thought then.

Also, it's entirely possible that I messed up during previous rebases when adapting the code to integrate with the speedwagon, so a more general audit/solution might be worth investigating as well.